// drizzle-orm@0.27.2/pg-core/index.cjs
'use strict';

var alias$1 = require('../alias-72a4082c.cjs');

function alias(table, alias) {
    return new Proxy(table, new alias$1.TableAliasProxyHandler(alias, false));
}

exports.Check = alias$1.Check;
exports.CheckBuilder = alias$1.CheckBuilder;
exports.DefaultViewBuilderCore = alias$1.DefaultViewBuilderCore;
exports.ForeignKey = alias$1.ForeignKey;
exports.ForeignKeyBuilder = alias$1.ForeignKeyBuilder;
exports.Index = alias$1.Index;
exports.IndexBuilder = alias$1.IndexBuilder;
exports.IndexBuilderOn = alias$1.IndexBuilderOn;
exports.InlineForeignKeys = alias$1.InlineForeignKeys;
exports.ManualMaterializedViewBuilder = alias$1.ManualMaterializedViewBuilder;
exports.ManualViewBuilder = alias$1.ManualViewBuilder;
exports.MaterializedViewBuilder = alias$1.MaterializedViewBuilder;
exports.MaterializedViewBuilderCore = alias$1.MaterializedViewBuilderCore;
exports.PgArray = alias$1.PgArray;
exports.PgArrayBuilder = alias$1.PgArrayBuilder;
exports.PgBigInt53 = alias$1.PgBigInt53;
exports.PgBigInt53Builder = alias$1.PgBigInt53Builder;
exports.PgBigInt64 = alias$1.PgBigInt64;
exports.PgBigInt64Builder = alias$1.PgBigInt64Builder;
exports.PgBigSerial53 = alias$1.PgBigSerial53;
exports.PgBigSerial53Builder = alias$1.PgBigSerial53Builder;
exports.PgBigSerial64 = alias$1.PgBigSerial64;
exports.PgBigSerial64Builder = alias$1.PgBigSerial64Builder;
exports.PgBoolean = alias$1.PgBoolean;
exports.PgBooleanBuilder = alias$1.PgBooleanBuilder;
exports.PgChar = alias$1.PgChar;
exports.PgCharBuilder = alias$1.PgCharBuilder;
exports.PgCidr = alias$1.PgCidr;
exports.PgCidrBuilder = alias$1.PgCidrBuilder;
exports.PgColumn = alias$1.PgColumn;
exports.PgColumnBuilder = alias$1.PgColumnBuilder;
exports.PgCustomColumn = alias$1.PgCustomColumn;
exports.PgCustomColumnBuilder = alias$1.PgCustomColumnBuilder;
exports.PgDatabase = alias$1.PgDatabase;
exports.PgDate = alias$1.PgDate;
exports.PgDateBuilder = alias$1.PgDateBuilder;
exports.PgDateString = alias$1.PgDateString;
exports.PgDateStringBuilder = alias$1.PgDateStringBuilder;
exports.PgDelete = alias$1.PgDelete;
exports.PgDialect = alias$1.PgDialect;
exports.PgDoublePrecision = alias$1.PgDoublePrecision;
exports.PgDoublePrecisionBuilder = alias$1.PgDoublePrecisionBuilder;
exports.PgEnumColumn = alias$1.PgEnumColumn;
exports.PgEnumColumnBuilder = alias$1.PgEnumColumnBuilder;
exports.PgInet = alias$1.PgInet;
exports.PgInetBuilder = alias$1.PgInetBuilder;
exports.PgInsert = alias$1.PgInsert;
exports.PgInsertBuilder = alias$1.PgInsertBuilder;
exports.PgInteger = alias$1.PgInteger;
exports.PgIntegerBuilder = alias$1.PgIntegerBuilder;
exports.PgInterval = alias$1.PgInterval;
exports.PgIntervalBuilder = alias$1.PgIntervalBuilder;
exports.PgJson = alias$1.PgJson;
exports.PgJsonBuilder = alias$1.PgJsonBuilder;
exports.PgJsonb = alias$1.PgJsonb;
exports.PgJsonbBuilder = alias$1.PgJsonbBuilder;
exports.PgMacaddr = alias$1.PgMacaddr;
exports.PgMacaddr8 = alias$1.PgMacaddr8;
exports.PgMacaddr8Builder = alias$1.PgMacaddr8Builder;
exports.PgMacaddrBuilder = alias$1.PgMacaddrBuilder;
exports.PgMaterializedView = alias$1.PgMaterializedView;
exports.PgMaterializedViewConfig = alias$1.PgMaterializedViewConfig;
exports.PgNumeric = alias$1.PgNumeric;
exports.PgNumericBuilder = alias$1.PgNumericBuilder;
exports.PgReal = alias$1.PgReal;
exports.PgRealBuilder = alias$1.PgRealBuilder;
exports.PgRefreshMaterializedView = alias$1.PgRefreshMaterializedView;
exports.PgSchema = alias$1.PgSchema;
exports.PgSelect = alias$1.PgSelect;
exports.PgSelectBuilder = alias$1.PgSelectBuilder;
exports.PgSelectQueryBuilder = alias$1.PgSelectQueryBuilder;
exports.PgSerial = alias$1.PgSerial;
exports.PgSerialBuilder = alias$1.PgSerialBuilder;
exports.PgSession = alias$1.PgSession;
exports.PgSmallInt = alias$1.PgSmallInt;
exports.PgSmallIntBuilder = alias$1.PgSmallIntBuilder;
exports.PgSmallSerial = alias$1.PgSmallSerial;
exports.PgSmallSerialBuilder = alias$1.PgSmallSerialBuilder;
exports.PgTable = alias$1.PgTable;
exports.PgText = alias$1.PgText;
exports.PgTextBuilder = alias$1.PgTextBuilder;
exports.PgTime = alias$1.PgTime;
exports.PgTimeBuilder = alias$1.PgTimeBuilder;
exports.PgTimestamp = alias$1.PgTimestamp;
exports.PgTimestampBuilder = alias$1.PgTimestampBuilder;
exports.PgTimestampString = alias$1.PgTimestampString;
exports.PgTimestampStringBuilder = alias$1.PgTimestampStringBuilder;
exports.PgTransaction = alias$1.PgTransaction;
exports.PgUUID = alias$1.PgUUID;
exports.PgUUIDBuilder = alias$1.PgUUIDBuilder;
exports.PgUpdate = alias$1.PgUpdate;
exports.PgUpdateBuilder = alias$1.PgUpdateBuilder;
exports.PgVarchar = alias$1.PgVarchar;
exports.PgVarcharBuilder = alias$1.PgVarcharBuilder;
exports.PgView = alias$1.PgView;
exports.PgViewBase = alias$1.PgViewBase;
exports.PgViewConfig = alias$1.PgViewConfig;
exports.PreparedQuery = alias$1.PreparedQuery;
exports.PrimaryKey = alias$1.PrimaryKey;
exports.PrimaryKeyBuilder = alias$1.PrimaryKeyBuilder;
exports.QueryBuilder = alias$1.QueryBuilder;
exports.UniqueConstraint = alias$1.UniqueConstraint;
exports.UniqueConstraintBuilder = alias$1.UniqueConstraintBuilder;
exports.UniqueOnConstraintBuilder = alias$1.UniqueOnConstraintBuilder;
exports.ViewBuilder = alias$1.ViewBuilder;
exports.bigint = alias$1.bigint;
exports.bigserial = alias$1.bigserial;
exports.boolean = alias$1.boolean;
exports.char = alias$1.char;
exports.check = alias$1.check;
exports.cidr = alias$1.cidr;
exports.customType = alias$1.customType;
exports.date = alias$1.date;
exports.decimal = alias$1.decimal;
exports.doublePrecision = alias$1.doublePrecision;
exports.foreignKey = alias$1.foreignKey;
exports.getMaterializedViewConfig = alias$1.getMaterializedViewConfig;
exports.getTableConfig = alias$1.getTableConfig;
exports.getViewConfig = alias$1.getViewConfig;
exports.index = alias$1.index;
exports.inet = alias$1.inet;
exports.integer = alias$1.integer;
exports.interval = alias$1.interval;
exports.isPgEnum = alias$1.isPgEnum;
exports.isPgSchema = alias$1.isPgSchema;
exports.json = alias$1.json;
exports.jsonb = alias$1.jsonb;
exports.macaddr = alias$1.macaddr;
exports.macaddr8 = alias$1.macaddr8;
exports.makePgArray = alias$1.makePgArray;
exports.numeric = alias$1.numeric;
exports.parsePgArray = alias$1.parsePgArray;
exports.parsePgNestedArray = alias$1.parsePgNestedArray;
exports.pgEnum = alias$1.pgEnum;
exports.pgMaterializedView = alias$1.pgMaterializedView;
exports.pgMaterializedViewWithSchema = alias$1.pgMaterializedViewWithSchema;
exports.pgSchema = alias$1.pgSchema;
exports.pgTable = alias$1.pgTable;
exports.pgTableCreator = alias$1.pgTableCreator;
exports.pgTableWithSchema = alias$1.pgTableWithSchema;
exports.pgView = alias$1.pgView;
exports.pgViewWithSchema = alias$1.pgViewWithSchema;
exports.primaryKey = alias$1.primaryKey;
exports.real = alias$1.real;
exports.serial = alias$1.serial;
exports.smallint = alias$1.smallint;
exports.smallserial = alias$1.smallserial;
exports.text = alias$1.text;
exports.time = alias$1.time;
exports.timestamp = alias$1.timestamp;
exports.unique = alias$1.unique;
exports.uniqueIndex = alias$1.uniqueIndex;
exports.uniqueKeyName = alias$1.uniqueKeyName;
exports.uuid = alias$1.uuid;
exports.varchar = alias$1.varchar;
exports.alias = alias;
