// kysely@0.26.1/dist/cjs/operation-node/delete-query-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteQueryNode = void 0;
const object_utils_js_1 = require("../util/object-utils.js");
const from_node_js_1 = require("./from-node.js");
const order_by_node_js_1 = require("./order-by-node.js");
const using_node_js_1 = require("./using-node.js");
/**
 * @internal
 */
exports.DeleteQueryNode = (0, object_utils_js_1.freeze)({
    is(node) {
        return node.kind === 'DeleteQueryNode';
    },
    create(fromItems, withNode) {
        return (0, object_utils_js_1.freeze)({
            kind: 'DeleteQueryNode',
            from: from_node_js_1.FromNode.create(fromItems),
            ...(withNode && { with: withNode }),
        });
    },
    cloneWithOrderByItem(deleteNode, item) {
        return (0, object_utils_js_1.freeze)({
            ...deleteNode,
            orderBy: deleteNode.orderBy
                ? order_by_node_js_1.OrderByNode.cloneWithItem(deleteNode.orderBy, item)
                : order_by_node_js_1.OrderByNode.create(item),
        });
    },
    cloneWithLimit(deleteNode, limit) {
        return (0, object_utils_js_1.freeze)({
            ...deleteNode,
            limit,
        });
    },
    cloneWithUsing(deleteNode, tables) {
        return (0, object_utils_js_1.freeze)({
            ...deleteNode,
            using: deleteNode.using !== undefined
                ? using_node_js_1.UsingNode.cloneWithTables(deleteNode.using, tables)
                : using_node_js_1.UsingNode.create(tables),
        });
    },
});
