// drizzle-orm@0.27.2/sqlite-core/index.mjs
import { M as TableAliasProxyHandler, e as entityKind, L as ColumnBuilder, C as Column, s as sql, T as Table, i as is, u as ViewBaseConfig } from '../alias-3e926a50.mjs';
import { d as SQLiteTable, e as SQLiteViewConfig } from '../session-8302855c.mjs';
export { B as BaseSQLiteDatabase, I as InlineForeignKeys, M as ManualViewBuilder, P as PreparedQuery, Q as QueryBuilder, c as SQLiteAsyncDialect, g as SQLiteDelete, f as SQLiteDialect, i as SQLiteInsert, h as SQLiteInsertBuilder, l as SQLiteSelect, j as SQLiteSelectBuilder, k as SQLiteSelectQueryBuilder, S as SQLiteSession, b as SQLiteSyncDialect, a as SQLiteTransaction, n as SQLiteUpdate, m as SQLiteUpdateBuilder, r as SQLiteView, q as SQLiteViewBase, p as ViewBuilder, V as ViewBuilderCore, s as sqliteTable, o as sqliteTableCreator, t as sqliteView, v as view } from '../session-8302855c.mjs';
import '../errors-fed11085.mjs';

function alias(table, alias) {
    return new Proxy(table, new TableAliasProxyHandler(alias, false));
}

class CheckBuilder {
    name;
    value;
    static [entityKind] = 'SQLiteCheckBuilder';
    brand;
    constructor(name, value) {
        this.name = name;
        this.value = value;
    }
    build(table) {
        return new Check(table, this);
    }
}
class Check {
    table;
    static [entityKind] = 'SQLiteCheck';
    name;
    value;
    constructor(table, builder) {
        this.table = table;
        this.name = builder.name;
        this.value = builder.value;
    }
}
function check(name, value) {
    return new CheckBuilder(name, value);
}

class ForeignKeyBuilder {
    static [entityKind] = 'SQLiteForeignKeyBuilder';
    /** @internal */
    reference;
    /** @internal */
    _onUpdate;
    /** @internal */
    _onDelete;
    constructor(config, actions) {
        this.reference = () => {
            const { columns, foreignColumns } = config();
            return { columns, foreignTable: foreignColumns[0].table, foreignColumns };
        };
        if (actions) {
            this._onUpdate = actions.onUpdate;
            this._onDelete = actions.onDelete;
        }
    }
    onUpdate(action) {
        this._onUpdate = action;
        return this;
    }
    onDelete(action) {
        this._onDelete = action;
        return this;
    }
    /** @internal */
    build(table) {
        return new ForeignKey(table, this);
    }
}
class ForeignKey {
    table;
    static [entityKind] = 'SQLiteForeignKey';
    reference;
    onUpdate;
    onDelete;
    constructor(table, builder) {
        this.table = table;
        this.reference = builder.reference;
        this.onUpdate = builder._onUpdate;
        this.onDelete = builder._onDelete;
    }
    getName() {
        const { columns, foreignColumns } = this.reference();
        const columnNames = columns.map((column) => column.name);
        const foreignColumnNames = foreignColumns.map((column) => column.name);
        const chunks = [
            this.table[SQLiteTable.Symbol.Name],
            ...columnNames,
            foreignColumns[0].table[SQLiteTable.Symbol.Name],
            ...foreignColumnNames,
        ];
        return `${chunks.join('_')}_fk`;
    }
}
function foreignKey(config) {
    function mappedConfig() {
        const { columns, foreignColumns } = config();
        return {
            columns,
            foreignColumns,
        };
    }
    return new ForeignKeyBuilder(mappedConfig);
}

function uniqueKeyName(table, columns) {
    return `${table[SQLiteTable.Symbol.Name]}_${columns.join('_')}_unique`;
}
function unique(name) {
    return new UniqueOnConstraintBuilder(name);
}
class UniqueConstraintBuilder {
    name;
    static [entityKind] = 'SQLiteUniqueConstraintBuilder';
    /** @internal */
    columns;
    constructor(columns, name) {
        this.name = name;
        this.columns = columns;
    }
    /** @internal */
    build(table) {
        return new UniqueConstraint(table, this.columns, this.name);
    }
}
class UniqueOnConstraintBuilder {
    static [entityKind] = 'SQLiteUniqueOnConstraintBuilder';
    /** @internal */
    name;
    constructor(name) {
        this.name = name;
    }
    on(...columns) {
        return new UniqueConstraintBuilder(columns, this.name);
    }
}
class UniqueConstraint {
    table;
    static [entityKind] = 'SQLiteUniqueConstraint';
    columns;
    name;
    constructor(table, columns, name) {
        this.table = table;
        this.columns = columns;
        this.name = name ?? uniqueKeyName(this.table, this.columns.map((column) => column.name));
    }
    getName() {
        return this.name;
    }
}

class SQLiteColumnBuilder extends ColumnBuilder {
    static [entityKind] = 'SQLiteColumnBuilder';
    foreignKeyConfigs = [];
    references(ref, actions = {}) {
        this.foreignKeyConfigs.push({ ref, actions });
        return this;
    }
    unique(name) {
        this.config.isUnique = true;
        this.config.uniqueName = name;
        return this;
    }
    /** @internal */
    buildForeignKeys(column, table) {
        return this.foreignKeyConfigs.map(({ ref, actions }) => {
            return ((ref, actions) => {
                const builder = new ForeignKeyBuilder(() => {
                    const foreignColumn = ref();
                    return { columns: [column], foreignColumns: [foreignColumn] };
                });
                if (actions.onUpdate) {
                    builder.onUpdate(actions.onUpdate);
                }
                if (actions.onDelete) {
                    builder.onDelete(actions.onDelete);
                }
                return builder.build(table);
            })(ref, actions);
        });
    }
}
// To understand how to use `SQLiteColumn` and `AnySQLiteColumn`, see `Column` and `AnyColumn` documentation.
class SQLiteColumn extends Column {
    table;
    static [entityKind] = 'SQLiteColumn';
    constructor(table, config) {
        if (!config.uniqueName) {
            config.uniqueName = uniqueKeyName(table, [config.name]);
        }
        super(table, config);
        this.table = table;
    }
}

class SQLiteBigIntBuilder extends SQLiteColumnBuilder {
    static [entityKind] = 'SQLiteBigIntBuilder';
    /** @internal */
    build(table) {
        return new SQLiteBigInt(table, this.config);
    }
}
class SQLiteBigInt extends SQLiteColumn {
    static [entityKind] = 'SQLiteBigInt';
    getSQLType() {
        return 'blob';
    }
    mapFromDriverValue(value) {
        return BigInt(value.toString());
    }
    mapToDriverValue(value) {
        return Buffer.from(value.toString());
    }
}
class SQLiteBlobJsonBuilder extends SQLiteColumnBuilder {
    static [entityKind] = 'SQLiteBlobJsonBuilder';
    /** @internal */
    build(table) {
        return new SQLiteBlobJson(table, this.config);
    }
}
class SQLiteBlobJson extends SQLiteColumn {
    static [entityKind] = 'SQLiteBlobJson';
    getSQLType() {
        return 'blob';
    }
    mapFromDriverValue(value) {
        return JSON.parse(value.toString());
    }
    mapToDriverValue(value) {
        return Buffer.from(JSON.stringify(value));
    }
}
class SQLiteBlobBufferBuilder extends SQLiteColumnBuilder {
    static [entityKind] = 'SQLiteBlobBufferBuilder';
    /** @internal */
    build(table) {
        return new SQLiteBlobBuffer(table, this.config);
    }
}
class SQLiteBlobBuffer extends SQLiteColumn {
    static [entityKind] = 'SQLiteBlobBuffer';
    getSQLType() {
        return 'blob';
    }
}
function blob(name, config) {
    if (config?.mode === 'json') {
        return new SQLiteBlobJsonBuilder(name);
    }
    if (config?.mode === 'bigint') {
        return new SQLiteBigIntBuilder(name);
    }
    return new SQLiteBlobBufferBuilder(name);
}

class SQLiteCustomColumnBuilder extends SQLiteColumnBuilder {
    static [entityKind] = 'SQLiteCustomColumnBuilder';
    constructor(name, fieldConfig, customTypeParams) {
        super(name);
        this.config.fieldConfig = fieldConfig;
        this.config.customTypeParams = customTypeParams;
    }
    /** @internal */
    build(table) {
        return new SQLiteCustomColumn(table, this.config);
    }
}
class SQLiteCustomColumn extends SQLiteColumn {
    static [entityKind] = 'SQLiteCustomColumn';
    sqlName;
    mapTo;
    mapFrom;
    constructor(table, config) {
        super(table, config);
        this.sqlName = config.customTypeParams.dataType(config.fieldConfig);
        this.mapTo = config.customTypeParams.toDriver;
        this.mapFrom = config.customTypeParams.fromDriver;
    }
    getSQLType() {
        return this.sqlName;
    }
    mapFromDriverValue(value) {
        return typeof this.mapFrom === 'function' ? this.mapFrom(value) : value;
    }
    mapToDriverValue(value) {
        return typeof this.mapTo === 'function' ? this.mapTo(value) : value;
    }
}
/**
 * Custom sqlite database data type generator
 */
function customType(customTypeParams) {
    return (dbName, fieldConfig) => {
        return new SQLiteCustomColumnBuilder(dbName, fieldConfig, customTypeParams);
    };
}

class SQLiteBaseIntegerBuilder extends SQLiteColumnBuilder {
    static [entityKind] = 'SQLiteBaseIntegerBuilder';
    constructor(name) {
        super(name);
        this.config.autoIncrement = false;
    }
    primaryKey(config) {
        if (config?.autoIncrement) {
            this.config.autoIncrement = true;
        }
        this.config.hasDefault = true;
        return super.primaryKey();
    }
}
class SQLiteBaseInteger extends SQLiteColumn {
    static [entityKind] = 'SQLiteBaseInteger';
    autoIncrement = this.config.autoIncrement;
    getSQLType() {
        return 'integer';
    }
}
class SQLiteIntegerBuilder extends SQLiteBaseIntegerBuilder {
    static [entityKind] = 'SQLiteIntegerBuilder';
    build(table) {
        return new SQLiteInteger(table, this.config);
    }
}
class SQLiteInteger extends SQLiteBaseInteger {
    static [entityKind] = 'SQLiteInteger';
}
class SQLiteTimestampBuilder extends SQLiteBaseIntegerBuilder {
    static [entityKind] = 'SQLiteTimestampBuilder';
    constructor(name, mode) {
        super(name);
        this.config.mode = mode;
    }
    /**
     * @deprecated Use `default()` with your own expression instead.
     *
     * Adds `DEFAULT (cast((julianday('now') - 2440587.5)*86400000 as integer))` to the column, which is the current epoch timestamp in milliseconds.
     */
    defaultNow() {
        return this.default(sql `(cast((julianday('now') - 2440587.5)*86400000 as integer))`);
    }
    build(table) {
        return new SQLiteTimestamp(table, this.config);
    }
}
class SQLiteTimestamp extends SQLiteBaseInteger {
    static [entityKind] = 'SQLiteTimestamp';
    mode = this.config.mode;
    mapFromDriverValue(value) {
        if (this.config.mode === 'timestamp') {
            return new Date(value * 1000);
        }
        return new Date(value);
    }
    mapToDriverValue(value) {
        const unix = value.getTime();
        if (this.config.mode === 'timestamp') {
            return Math.floor(unix / 1000);
        }
        return unix;
    }
}
class SQLiteBooleanBuilder extends SQLiteBaseIntegerBuilder {
    static [entityKind] = 'SQLiteBooleanBuilder';
    constructor(name, mode) {
        super(name);
        this.config.mode = mode;
    }
    build(table) {
        return new SQLiteBoolean(table, this.config);
    }
}
class SQLiteBoolean extends SQLiteBaseInteger {
    static [entityKind] = 'SQLiteBoolean';
    mode = this.config.mode;
    mapFromDriverValue(value) {
        return Number(value) === 1;
    }
    mapToDriverValue(value) {
        return value ? 1 : 0;
    }
}
function integer(name, config) {
    if (config?.mode === 'timestamp' || config?.mode === 'timestamp_ms') {
        return new SQLiteTimestampBuilder(name, config.mode);
    }
    if (config?.mode === 'boolean') {
        return new SQLiteBooleanBuilder(name, config.mode);
    }
    return new SQLiteIntegerBuilder(name);
}
const int = integer;

class SQLiteNumericBuilder extends SQLiteColumnBuilder {
    static [entityKind] = 'SQLiteNumericBuilder';
    /** @internal */
    build(table) {
        return new SQLiteNumeric(table, this.config);
    }
}
class SQLiteNumeric extends SQLiteColumn {
    static [entityKind] = 'SQLiteNumeric';
    getSQLType() {
        return 'numeric';
    }
}
function numeric(name) {
    return new SQLiteNumericBuilder(name);
}

class SQLiteRealBuilder extends SQLiteColumnBuilder {
    static [entityKind] = 'SQLiteRealBuilder';
    /** @internal */
    build(table) {
        return new SQLiteReal(table, this.config);
    }
}
class SQLiteReal extends SQLiteColumn {
    static [entityKind] = 'SQLiteReal';
    getSQLType() {
        return 'real';
    }
}
function real(name) {
    return new SQLiteRealBuilder(name);
}

class SQLiteTextBuilder extends SQLiteColumnBuilder {
    static [entityKind] = 'SQLiteTextBuilder';
    constructor(name, config) {
        super(name);
        this.config.enumValues = (config.enum ?? []);
        this.config.length = config.length;
    }
    /** @internal */
    build(table) {
        return new SQLiteText(table, this.config);
    }
}
class SQLiteText extends SQLiteColumn {
    static [entityKind] = 'SQLiteText';
    enumValues = this.config.enumValues;
    length = this.config.length;
    constructor(table, config) {
        super(table, config);
    }
    getSQLType() {
        return `text${this.config.length ? `(${this.config.length})` : ''}`;
    }
}
function text(name, config = {}) {
    return new SQLiteTextBuilder(name, config);
}

class IndexBuilderOn {
    name;
    unique;
    static [entityKind] = 'SQLiteIndexBuilderOn';
    constructor(name, unique) {
        this.name = name;
        this.unique = unique;
    }
    on(...columns) {
        return new IndexBuilder(this.name, columns, this.unique);
    }
}
class IndexBuilder {
    static [entityKind] = 'SQLiteIndexBuilder';
    /** @internal */
    config;
    constructor(name, columns, unique) {
        this.config = {
            name,
            columns,
            unique,
            where: undefined,
        };
    }
    /**
     * Condition for partial index.
     */
    where(condition) {
        this.config.where = condition;
        return this;
    }
    /** @internal */
    build(table) {
        return new Index(this.config, table);
    }
}
class Index {
    static [entityKind] = 'SQLiteIndex';
    config;
    constructor(config, table) {
        this.config = { ...config, table };
    }
}
function index(name) {
    return new IndexBuilderOn(name, false);
}
function uniqueIndex(name) {
    return new IndexBuilderOn(name, true);
}

function primaryKey(...columns) {
    return new PrimaryKeyBuilder(columns);
}
class PrimaryKeyBuilder {
    static [entityKind] = 'SQLitePrimaryKeyBuilder';
    /** @internal */
    columns;
    constructor(columns) {
        this.columns = columns;
    }
    /** @internal */
    build(table) {
        return new PrimaryKey(table, this.columns);
    }
}
class PrimaryKey {
    table;
    static [entityKind] = 'SQLitePrimaryKey';
    columns;
    constructor(table, columns) {
        this.table = table;
        this.columns = columns;
    }
    getName() {
        return `${this.table[SQLiteTable.Symbol.Name]}_${this.columns.map((column) => column.name).join('_')}_pk`;
    }
}

function getTableConfig(table) {
    const columns = Object.values(table[SQLiteTable.Symbol.Columns]);
    const indexes = [];
    const checks = [];
    const primaryKeys = [];
    const uniqueConstraints = [];
    const foreignKeys = Object.values(table[SQLiteTable.Symbol.InlineForeignKeys]);
    const name = table[Table.Symbol.Name];
    const extraConfigBuilder = table[SQLiteTable.Symbol.ExtraConfigBuilder];
    if (extraConfigBuilder !== undefined) {
        const extraConfig = extraConfigBuilder(table[SQLiteTable.Symbol.Columns]);
        for (const builder of Object.values(extraConfig)) {
            if (is(builder, IndexBuilder)) {
                indexes.push(builder.build(table));
            }
            else if (is(builder, CheckBuilder)) {
                checks.push(builder.build(table));
            }
            else if (is(builder, UniqueConstraintBuilder)) {
                uniqueConstraints.push(builder.build(table));
            }
            else if (is(builder, PrimaryKeyBuilder)) {
                primaryKeys.push(builder.build(table));
            }
            else if (is(builder, ForeignKeyBuilder)) {
                foreignKeys.push(builder.build(table));
            }
        }
    }
    return {
        columns,
        indexes,
        foreignKeys,
        checks,
        primaryKeys,
        uniqueConstraints,
        name,
    };
}
function getViewConfig(view) {
    return {
        ...view[ViewBaseConfig],
        ...view[SQLiteViewConfig],
    };
}

export { Check, CheckBuilder, ForeignKey, ForeignKeyBuilder, Index, IndexBuilder, IndexBuilderOn, PrimaryKey, PrimaryKeyBuilder, SQLiteBaseInteger, SQLiteBaseIntegerBuilder, SQLiteBigInt, SQLiteBigIntBuilder, SQLiteBlobBuffer, SQLiteBlobBufferBuilder, SQLiteBlobJson, SQLiteBlobJsonBuilder, SQLiteBoolean, SQLiteBooleanBuilder, SQLiteColumn, SQLiteColumnBuilder, SQLiteCustomColumn, SQLiteCustomColumnBuilder, SQLiteInteger, SQLiteIntegerBuilder, SQLiteNumeric, SQLiteNumericBuilder, SQLiteReal, SQLiteRealBuilder, SQLiteTable, SQLiteText, SQLiteTextBuilder, SQLiteTimestamp, SQLiteTimestampBuilder, SQLiteViewConfig, UniqueConstraint, UniqueConstraintBuilder, UniqueOnConstraintBuilder, alias, blob, check, customType, foreignKey, getTableConfig, getViewConfig, index, int, integer, numeric, primaryKey, real, text, unique, uniqueIndex, uniqueKeyName };
