// kysely@0.26.1/dist/cjs/expression/expression-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.expressionBuilder = exports.createExpressionBuilder = void 0;
const select_query_builder_js_1 = require("../query-builder/select-query-builder.js");
const select_query_node_js_1 = require("../operation-node/select-query-node.js");
const table_parser_js_1 = require("../parser/table-parser.js");
const with_schema_plugin_js_1 = require("../plugin/with-schema/with-schema-plugin.js");
const query_id_js_1 = require("../util/query-id.js");
const function_module_js_1 = require("../query-builder/function-module.js");
const reference_parser_js_1 = require("../parser/reference-parser.js");
const binary_operation_parser_js_1 = require("../parser/binary-operation-parser.js");
const and_node_js_1 = require("../operation-node/and-node.js");
const or_node_js_1 = require("../operation-node/or-node.js");
const parens_node_js_1 = require("../operation-node/parens-node.js");
const expression_wrapper_js_1 = require("./expression-wrapper.js");
const unary_operation_parser_js_1 = require("../parser/unary-operation-parser.js");
const value_parser_js_1 = require("../parser/value-parser.js");
const noop_query_executor_js_1 = require("../query-executor/noop-query-executor.js");
const value_node_js_1 = require("../operation-node/value-node.js");
const case_builder_js_1 = require("../query-builder/case-builder.js");
const case_node_js_1 = require("../operation-node/case-node.js");
const object_utils_js_1 = require("../util/object-utils.js");
const json_path_builder_js_1 = require("../query-builder/json-path-builder.js");
function createExpressionBuilder(executor = noop_query_executor_js_1.NOOP_QUERY_EXECUTOR) {
    function binary(lhs, op, rhs) {
        return new expression_wrapper_js_1.ExpressionWrapper((0, binary_operation_parser_js_1.parseValueBinaryOperation)(lhs, op, rhs));
    }
    function unary(op, expr) {
        return new expression_wrapper_js_1.ExpressionWrapper((0, unary_operation_parser_js_1.parseUnaryOperation)(op, expr));
    }
    const eb = Object.assign(binary, {
        fn: undefined,
        eb: undefined,
        selectFrom(table) {
            return new select_query_builder_js_1.SelectQueryBuilder({
                queryId: (0, query_id_js_1.createQueryId)(),
                executor: executor,
                queryNode: select_query_node_js_1.SelectQueryNode.create((0, table_parser_js_1.parseTableExpressionOrList)(table)),
            });
        },
        case(reference) {
            return new case_builder_js_1.CaseBuilder({
                node: case_node_js_1.CaseNode.create((0, object_utils_js_1.isUndefined)(reference)
                    ? undefined
                    : (0, reference_parser_js_1.parseReferenceExpression)(reference)),
            });
        },
        ref(reference, op) {
            if ((0, object_utils_js_1.isUndefined)(op)) {
                return new expression_wrapper_js_1.ExpressionWrapper((0, reference_parser_js_1.parseStringReference)(reference));
            }
            return new json_path_builder_js_1.JSONPathBuilder((0, reference_parser_js_1.parseJSONReference)(reference, op));
        },
        val(value) {
            return new expression_wrapper_js_1.ExpressionWrapper((0, value_parser_js_1.parseValueExpressionOrList)(value));
        },
        // @deprecated
        cmpr(lhs, op, rhs) {
            return new expression_wrapper_js_1.ExpressionWrapper((0, binary_operation_parser_js_1.parseValueBinaryOperation)(lhs, op, rhs));
        },
        // @deprecated
        bxp(lhs, op, rhs) {
            return new expression_wrapper_js_1.ExpressionWrapper((0, binary_operation_parser_js_1.parseValueBinaryOperation)(lhs, op, rhs));
        },
        unary,
        not(expr) {
            return unary('not', expr);
        },
        exists(expr) {
            return unary('exists', expr);
        },
        neg(expr) {
            return unary('-', expr);
        },
        and(exprs) {
            if (exprs.length === 0) {
                return new expression_wrapper_js_1.ExpressionWrapper(value_node_js_1.ValueNode.createImmediate(true));
            }
            else if (exprs.length === 1) {
                return new expression_wrapper_js_1.ExpressionWrapper(exprs[0].toOperationNode());
            }
            let node = and_node_js_1.AndNode.create(exprs[0].toOperationNode(), exprs[1].toOperationNode());
            for (let i = 2; i < exprs.length; ++i) {
                node = and_node_js_1.AndNode.create(node, exprs[i].toOperationNode());
            }
            return new expression_wrapper_js_1.ExpressionWrapper(parens_node_js_1.ParensNode.create(node));
        },
        or(exprs) {
            if (exprs.length === 0) {
                return new expression_wrapper_js_1.ExpressionWrapper(value_node_js_1.ValueNode.createImmediate(false));
            }
            else if (exprs.length === 1) {
                return new expression_wrapper_js_1.ExpressionWrapper(exprs[0].toOperationNode());
            }
            let node = or_node_js_1.OrNode.create(exprs[0].toOperationNode(), exprs[1].toOperationNode());
            for (let i = 2; i < exprs.length; ++i) {
                node = or_node_js_1.OrNode.create(node, exprs[i].toOperationNode());
            }
            return new expression_wrapper_js_1.ExpressionWrapper(parens_node_js_1.ParensNode.create(node));
        },
        parens(...args) {
            const node = (0, binary_operation_parser_js_1.parseValueBinaryOperationOrExpression)(args);
            if (parens_node_js_1.ParensNode.is(node)) {
                // No double wrapping.
                return new expression_wrapper_js_1.ExpressionWrapper(node);
            }
            else {
                return new expression_wrapper_js_1.ExpressionWrapper(parens_node_js_1.ParensNode.create(node));
            }
        },
        withSchema(schema) {
            return createExpressionBuilder(executor.withPluginAtFront(new with_schema_plugin_js_1.WithSchemaPlugin(schema)));
        },
    });
    eb.fn = (0, function_module_js_1.createFunctionModule)();
    eb.eb = eb;
    return eb;
}
exports.createExpressionBuilder = createExpressionBuilder;
function expressionBuilder(_) {
    return createExpressionBuilder();
}
exports.expressionBuilder = expressionBuilder;
