// drizzle-orm@0.27.2/mysql-core/index.cjs
'use strict';

var alias$1 = require('../alias-72a4082c.cjs');
var session = require('../session-5f9eaf6b.cjs');

function alias(table, alias) {
    return new Proxy(table, new alias$1.TableAliasProxyHandler(alias, false));
}

var _a$s, _b$r;
class CheckBuilder {
    constructor(name, value) {
        this.name = name;
        this.value = value;
    }
    /** @internal */
    build(table) {
        return new Check(table, this);
    }
}
_a$s = alias$1.entityKind;
CheckBuilder[_a$s] = 'MySqlCheckBuilder';
class Check {
    constructor(table, builder) {
        this.table = table;
        this.name = builder.name;
        this.value = builder.value;
    }
}
_b$r = alias$1.entityKind;
Check[_b$r] = 'MySqlCheck';
function check(name, value) {
    return new CheckBuilder(name, value);
}

var _a$r, _b$q, _c$4, _d$3;
class MySqlBigInt53Builder extends session.MySqlColumnBuilderWithAutoIncrement {
    /** @internal */
    build(table) {
        return new MySqlBigInt53(table, this.config);
    }
}
_a$r = alias$1.entityKind;
MySqlBigInt53Builder[_a$r] = 'MySqlBigInt53Builder';
class MySqlBigInt53 extends session.MySqlColumnWithAutoIncrement {
    getSQLType() {
        return 'bigint';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'number') {
            return value;
        }
        return Number(value);
    }
}
_b$q = alias$1.entityKind;
MySqlBigInt53[_b$q] = 'MySqlBigInt53';
class MySqlBigInt64Builder extends session.MySqlColumnBuilderWithAutoIncrement {
    /** @internal */
    build(table) {
        return new MySqlBigInt64(table, this.config);
    }
}
_c$4 = alias$1.entityKind;
MySqlBigInt64Builder[_c$4] = 'MySqlBigInt64Builder';
class MySqlBigInt64 extends session.MySqlColumnWithAutoIncrement {
    getSQLType() {
        return 'bigint';
    }
    // eslint-disable-next-line unicorn/prefer-native-coercion-functions
    mapFromDriverValue(value) {
        return BigInt(value);
    }
}
_d$3 = alias$1.entityKind;
MySqlBigInt64[_d$3] = 'MySqlBigInt64';
function bigint(name, config) {
    if (config.mode === 'number') {
        return new MySqlBigInt53Builder(name);
    }
    return new MySqlBigInt64Builder(name);
}

var _a$q, _b$p;
class MySqlBinaryBuilder extends session.MySqlColumnBuilder {
    constructor(name, length) {
        super(name);
        this.config.length = length;
    }
    /** @internal */
    build(table) {
        return new MySqlBinary(table, this.config);
    }
}
_a$q = alias$1.entityKind;
MySqlBinaryBuilder[_a$q] = 'MySqlBinaryBuilder';
class MySqlBinary extends session.MySqlColumn {
    constructor() {
        super(...arguments);
        this.length = this.config.length;
    }
    getSQLType() {
        return this.length === undefined ? `binary` : `binary(${this.length})`;
    }
}
_b$p = alias$1.entityKind;
MySqlBinary[_b$p] = 'MySqlBinary';
function binary(name, config = {}) {
    return new MySqlBinaryBuilder(name, config.length);
}

var _a$p, _b$o;
class MySqlBooleanBuilder extends session.MySqlColumnBuilder {
    /** @internal */
    build(table) {
        return new MySqlBoolean(table, this.config);
    }
}
_a$p = alias$1.entityKind;
MySqlBooleanBuilder[_a$p] = 'MySqlBooleanBuilder';
class MySqlBoolean extends session.MySqlColumn {
    getSQLType() {
        return 'boolean';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'boolean') {
            return value;
        }
        return value === 1;
    }
}
_b$o = alias$1.entityKind;
MySqlBoolean[_b$o] = 'MySqlBoolean';
function boolean(name) {
    return new MySqlBooleanBuilder(name);
}

var _a$o, _b$n;
class MySqlCharBuilder extends session.MySqlColumnBuilder {
    constructor(name, config) {
        super(name);
        this.config.length = config.length;
        this.config.enum = config.enum;
    }
    /** @internal */
    build(table) {
        return new MySqlChar(table, this.config);
    }
}
_a$o = alias$1.entityKind;
MySqlCharBuilder[_a$o] = 'MySqlCharBuilder';
class MySqlChar extends session.MySqlColumn {
    constructor() {
        super(...arguments);
        this.length = this.config.length;
        this.enumValues = (this.config.enum ?? []);
    }
    getSQLType() {
        return this.length === undefined ? `char` : `char(${this.length})`;
    }
}
_b$n = alias$1.entityKind;
MySqlChar[_b$n] = 'MySqlChar';
function char(name, config = {}) {
    return new MySqlCharBuilder(name, config);
}

var _a$n, _b$m;
class MySqlCustomColumnBuilder extends session.MySqlColumnBuilder {
    constructor(name, fieldConfig, customTypeParams) {
        super(name);
        this.config.fieldConfig = fieldConfig;
        this.config.customTypeParams = customTypeParams;
    }
    /** @internal */
    build(table) {
        return new MySqlCustomColumn(table, this.config);
    }
}
_a$n = alias$1.entityKind;
MySqlCustomColumnBuilder[_a$n] = 'MySqlCustomColumnBuilder';
class MySqlCustomColumn extends session.MySqlColumn {
    constructor(table, config) {
        super(table, config);
        this.sqlName = config.customTypeParams.dataType(config.fieldConfig);
        this.mapTo = config.customTypeParams.toDriver;
        this.mapFrom = config.customTypeParams.fromDriver;
    }
    getSQLType() {
        return this.sqlName;
    }
    mapFromDriverValue(value) {
        return typeof this.mapFrom === 'function' ? this.mapFrom(value) : value;
    }
    mapToDriverValue(value) {
        return typeof this.mapTo === 'function' ? this.mapTo(value) : value;
    }
}
_b$m = alias$1.entityKind;
MySqlCustomColumn[_b$m] = 'MySqlCustomColumn';
/**
 * Custom mysql database data type generator
 */
function customType(customTypeParams) {
    return (dbName, fieldConfig) => {
        return new MySqlCustomColumnBuilder(dbName, fieldConfig, customTypeParams);
    };
}

var _a$m, _b$l, _c$3, _d$2;
class MySqlDateBuilder extends session.MySqlColumnBuilder {
    /** @internal */
    build(table) {
        return new MySqlDate(table, this.config);
    }
}
_a$m = alias$1.entityKind;
MySqlDateBuilder[_a$m] = 'MySqlDateBuilder';
class MySqlDate extends session.MySqlColumn {
    constructor(table, config) {
        super(table, config);
    }
    getSQLType() {
        return `date`;
    }
    mapFromDriverValue(value) {
        return new Date(value);
    }
}
_b$l = alias$1.entityKind;
MySqlDate[_b$l] = 'MySqlDate';
class MySqlDateStringBuilder extends session.MySqlColumnBuilder {
    /** @internal */
    build(table) {
        return new MySqlDateString(table, this.config);
    }
}
_c$3 = alias$1.entityKind;
MySqlDateStringBuilder[_c$3] = 'MySqlDateStringBuilder';
class MySqlDateString extends session.MySqlColumn {
    constructor(table, config) {
        super(table, config);
    }
    getSQLType() {
        return `date`;
    }
}
_d$2 = alias$1.entityKind;
MySqlDateString[_d$2] = 'MySqlDateString';
function date(name, config = {}) {
    if (config.mode === 'string') {
        return new MySqlDateStringBuilder(name);
    }
    return new MySqlDateBuilder(name);
}

var _a$l, _b$k, _c$2, _d$1;
class MySqlDateTimeBuilder extends session.MySqlColumnBuilder {
    constructor(name, config) {
        super(name);
        this.config.fsp = config?.fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlDateTime(table, this.config);
    }
}
_a$l = alias$1.entityKind;
MySqlDateTimeBuilder[_a$l] = 'MySqlDateTimeBuilder';
class MySqlDateTime extends session.MySqlColumn {
    constructor(table, config) {
        super(table, config);
        this.fsp = config.fsp;
    }
    getSQLType() {
        const precision = this.fsp === undefined ? '' : `(${this.fsp})`;
        return `datetime${precision}`;
    }
    mapFromDriverValue(value) {
        return new Date(value);
    }
}
_b$k = alias$1.entityKind;
MySqlDateTime[_b$k] = 'MySqlDateTime';
class MySqlDateTimeStringBuilder extends session.MySqlColumnBuilder {
    constructor(name, config) {
        super(name);
        this.config.fsp = config?.fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlDateTimeString(table, this.config);
    }
}
_c$2 = alias$1.entityKind;
MySqlDateTimeStringBuilder[_c$2] = 'MySqlDateTimeStringBuilder';
class MySqlDateTimeString extends session.MySqlColumn {
    constructor(table, config) {
        super(table, config);
        this.fsp = config.fsp;
    }
    getSQLType() {
        const precision = this.fsp === undefined ? '' : `(${this.fsp})`;
        return `datetime${precision}`;
    }
}
_d$1 = alias$1.entityKind;
MySqlDateTimeString[_d$1] = 'MySqlDateTimeString';
function datetime(name, config = {}) {
    if (config.mode === 'string') {
        return new MySqlDateTimeStringBuilder(name, config);
    }
    return new MySqlDateTimeBuilder(name, config);
}

var _a$k, _b$j;
class MySqlDecimalBuilder extends session.MySqlColumnBuilderWithAutoIncrement {
    constructor(name, precision, scale) {
        super(name);
        this.config.precision = precision;
        this.config.scale = scale;
    }
    /** @internal */
    build(table) {
        return new MySqlDecimal(table, this.config);
    }
}
_a$k = alias$1.entityKind;
MySqlDecimalBuilder[_a$k] = 'MySqlDecimalBuilder';
class MySqlDecimal extends session.MySqlColumnWithAutoIncrement {
    constructor() {
        super(...arguments);
        this.precision = this.config.precision;
        this.scale = this.config.scale;
    }
    getSQLType() {
        if (this.precision !== undefined && this.scale !== undefined) {
            return `decimal(${this.precision},${this.scale})`;
        }
        else if (this.precision === undefined) {
            return 'decimal';
        }
        else {
            return `decimal(${this.precision})`;
        }
    }
}
_b$j = alias$1.entityKind;
MySqlDecimal[_b$j] = 'MySqlDecimal';
function decimal(name, config = {}) {
    return new MySqlDecimalBuilder(name, config.precision, config.scale);
}

var _a$j, _b$i;
class MySqlDoubleBuilder extends session.MySqlColumnBuilderWithAutoIncrement {
    constructor(name, config) {
        super(name);
        this.config.precision = config?.precision;
        this.config.scale = config?.scale;
    }
    /** @internal */
    build(table) {
        return new MySqlDouble(table, this.config);
    }
}
_a$j = alias$1.entityKind;
MySqlDoubleBuilder[_a$j] = 'MySqlDoubleBuilder';
class MySqlDouble extends session.MySqlColumnWithAutoIncrement {
    constructor() {
        super(...arguments);
        this.precision = this.config.precision;
        this.scale = this.config.scale;
    }
    getSQLType() {
        if (this.precision !== undefined && this.scale !== undefined) {
            return `double(${this.precision},${this.scale})`;
        }
        else if (this.precision === undefined) {
            return 'double';
        }
        else {
            return `double(${this.precision})`;
        }
    }
}
_b$i = alias$1.entityKind;
MySqlDouble[_b$i] = 'MySqlDouble';
function double(name, config) {
    return new MySqlDoubleBuilder(name, config);
}

var _a$i, _b$h;
class MySqlEnumColumnBuilder extends session.MySqlColumnBuilder {
    constructor(name, values) {
        super(name);
        this.config.enumValues = values;
    }
    /** @internal */
    build(table) {
        return new MySqlEnumColumn(table, this.config);
    }
}
_a$i = alias$1.entityKind;
MySqlEnumColumnBuilder[_a$i] = 'MySqlEnumColumnBuilder';
class MySqlEnumColumn extends session.MySqlColumn {
    constructor() {
        super(...arguments);
        this.enumValues = this.config.enumValues;
    }
    getSQLType() {
        return `enum(${this.enumValues.map((value) => `'${value}'`).join(',')})`;
    }
}
_b$h = alias$1.entityKind;
MySqlEnumColumn[_b$h] = 'MySqlEnumColumn';
function mysqlEnum(name, values) {
    if (values.length === 0) {
        throw new Error(`You have an empty array for "${name}" enum values`);
    }
    return new MySqlEnumColumnBuilder(name, values);
}

var _a$h, _b$g;
class MySqlFloatBuilder extends session.MySqlColumnBuilderWithAutoIncrement {
    /** @internal */
    build(table) {
        return new MySqlFloat(table, this.config);
    }
}
_a$h = alias$1.entityKind;
MySqlFloatBuilder[_a$h] = 'MySqlFloatBuilder';
class MySqlFloat extends session.MySqlColumnWithAutoIncrement {
    getSQLType() {
        return 'float';
    }
}
_b$g = alias$1.entityKind;
MySqlFloat[_b$g] = 'MySqlFloat';
function float(name) {
    return new MySqlFloatBuilder(name);
}

var _a$g, _b$f;
class MySqlIntBuilder extends session.MySqlColumnBuilderWithAutoIncrement {
    /** @internal */
    build(table) {
        return new MySqlInt(table, this.config);
    }
}
_a$g = alias$1.entityKind;
MySqlIntBuilder[_a$g] = 'MySqlIntBuilder';
class MySqlInt extends session.MySqlColumnWithAutoIncrement {
    getSQLType() {
        return 'int';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'string') {
            return Number(value);
        }
        return value;
    }
}
_b$f = alias$1.entityKind;
MySqlInt[_b$f] = 'MySqlInt';
function int(name) {
    return new MySqlIntBuilder(name);
}

var _a$f, _b$e;
class MySqlJsonBuilder extends session.MySqlColumnBuilder {
    /** @internal */
    build(table) {
        return new MySqlJson(table, this.config);
    }
}
_a$f = alias$1.entityKind;
MySqlJsonBuilder[_a$f] = 'MySqlJsonBuilder';
class MySqlJson extends session.MySqlColumn {
    getSQLType() {
        return 'json';
    }
    mapToDriverValue(value) {
        return JSON.stringify(value);
    }
}
_b$e = alias$1.entityKind;
MySqlJson[_b$e] = 'MySqlJson';
function json(name) {
    return new MySqlJsonBuilder(name);
}

var _a$e, _b$d;
class MySqlMediumIntBuilder extends session.MySqlColumnBuilderWithAutoIncrement {
    /** @internal */
    build(table) {
        return new MySqlMediumInt(table, this.config);
    }
}
_a$e = alias$1.entityKind;
MySqlMediumIntBuilder[_a$e] = 'MySqlMediumIntBuilder';
class MySqlMediumInt extends session.MySqlColumnWithAutoIncrement {
    getSQLType() {
        return 'mediumint';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'string') {
            return Number(value);
        }
        return value;
    }
}
_b$d = alias$1.entityKind;
MySqlMediumInt[_b$d] = 'MySqlMediumInt';
function mediumint(name) {
    return new MySqlMediumIntBuilder(name);
}

var _a$d, _b$c;
class MySqlRealBuilder extends session.MySqlColumnBuilderWithAutoIncrement {
    constructor(name, config) {
        super(name);
        this.config.precision = config?.precision;
        this.config.scale = config?.scale;
    }
    /** @internal */
    build(table) {
        return new MySqlReal(table, this.config);
    }
}
_a$d = alias$1.entityKind;
MySqlRealBuilder[_a$d] = 'MySqlRealBuilder';
class MySqlReal extends session.MySqlColumnWithAutoIncrement {
    constructor() {
        super(...arguments);
        this.precision = this.config.precision;
        this.scale = this.config.scale;
    }
    getSQLType() {
        if (this.precision !== undefined && this.scale !== undefined) {
            return `real(${this.precision}, ${this.scale})`;
        }
        else if (this.precision === undefined) {
            return 'real';
        }
        else {
            return `real(${this.precision})`;
        }
    }
}
_b$c = alias$1.entityKind;
MySqlReal[_b$c] = 'MySqlReal';
function real(name, config = {}) {
    return new MySqlRealBuilder(name, config);
}

var _a$c, _b$b;
class MySqlSerialBuilder extends session.MySqlColumnBuilderWithAutoIncrement {
    constructor(name) {
        super(name);
        this.config.hasDefault = true;
        this.config.autoIncrement = true;
    }
    /** @internal */
    build(table) {
        return new MySqlSerial(table, this.config);
    }
}
_a$c = alias$1.entityKind;
MySqlSerialBuilder[_a$c] = 'MySqlSerialBuilder';
class MySqlSerial extends session.MySqlColumnWithAutoIncrement {
    getSQLType() {
        return 'serial';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'string') {
            return Number(value);
        }
        return value;
    }
}
_b$b = alias$1.entityKind;
MySqlSerial[_b$b] = 'MySqlSerial';
function serial(name) {
    return new MySqlSerialBuilder(name);
}

var _a$b, _b$a;
class MySqlSmallIntBuilder extends session.MySqlColumnBuilderWithAutoIncrement {
    /** @internal */
    build(table) {
        return new MySqlSmallInt(table, this.config);
    }
}
_a$b = alias$1.entityKind;
MySqlSmallIntBuilder[_a$b] = 'MySqlSmallIntBuilder';
class MySqlSmallInt extends session.MySqlColumnWithAutoIncrement {
    getSQLType() {
        return 'smallint';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'string') {
            return Number(value);
        }
        return value;
    }
}
_b$a = alias$1.entityKind;
MySqlSmallInt[_b$a] = 'MySqlSmallInt';
function smallint(name) {
    return new MySqlSmallIntBuilder(name);
}

var _a$a, _b$9;
class MySqlTextBuilder extends session.MySqlColumnBuilder {
    constructor(name, textType, config) {
        super(name);
        this.config.textType = textType;
        this.config.enumValues = config.enum;
    }
    /** @internal */
    build(table) {
        return new MySqlText(table, this.config);
    }
}
_a$a = alias$1.entityKind;
MySqlTextBuilder[_a$a] = 'MySqlTextBuilder';
class MySqlText extends session.MySqlColumn {
    constructor() {
        super(...arguments);
        this.textType = this.config.textType;
        this.enumValues = (this.config.enumValues ?? []);
    }
    getSQLType() {
        return this.textType;
    }
}
_b$9 = alias$1.entityKind;
MySqlText[_b$9] = 'MySqlText';
function text(name, config = {}) {
    return new MySqlTextBuilder(name, 'text', config);
}
function tinytext(name, config = {}) {
    return new MySqlTextBuilder(name, 'tinytext', config);
}
function mediumtext(name, config = {}) {
    return new MySqlTextBuilder(name, 'mediumtext', config);
}
function longtext(name, config = {}) {
    return new MySqlTextBuilder(name, 'longtext', config);
}

var _a$9, _b$8;
class MySqlTimeBuilder extends session.MySqlColumnBuilder {
    constructor(name, config) {
        super(name);
        this.config.fsp = config?.fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlTime(table, this.config);
    }
}
_a$9 = alias$1.entityKind;
MySqlTimeBuilder[_a$9] = 'MySqlTimeBuilder';
class MySqlTime extends session.MySqlColumn {
    constructor() {
        super(...arguments);
        this.fsp = this.config.fsp;
    }
    getSQLType() {
        const precision = this.fsp === undefined ? '' : `(${this.fsp})`;
        return `time${precision}`;
    }
}
_b$8 = alias$1.entityKind;
MySqlTime[_b$8] = 'MySqlTime';
function time(name, config) {
    return new MySqlTimeBuilder(name, config);
}

var _a$8, _b$7;
class MySqlDateColumnBaseBuilder extends session.MySqlColumnBuilder {
    defaultNow() {
        return this.default(alias$1.sql `(now())`);
    }
    // "on update now" also adds an implicit default value to the column - https://dev.mysql.com/doc/refman/8.0/en/timestamp-initialization.html
    onUpdateNow() {
        this.config.hasOnUpdateNow = true;
        this.config.hasDefault = true;
        return this;
    }
}
_a$8 = alias$1.entityKind;
MySqlDateColumnBaseBuilder[_a$8] = 'MySqlDateColumnBuilder';
class MySqlDateBaseColumn extends session.MySqlColumn {
    constructor() {
        super(...arguments);
        this.hasOnUpdateNow = this.config.hasOnUpdateNow;
    }
}
_b$7 = alias$1.entityKind;
MySqlDateBaseColumn[_b$7] = 'MySqlDateColumn';

var _a$7, _b$6, _c$1, _d;
class MySqlTimestampBuilder extends MySqlDateColumnBaseBuilder {
    constructor(name, config) {
        super(name);
        this.config.fsp = config?.fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlTimestamp(table, this.config);
    }
}
_a$7 = alias$1.entityKind;
MySqlTimestampBuilder[_a$7] = 'MySqlTimestampBuilder';
class MySqlTimestamp extends MySqlDateBaseColumn {
    constructor() {
        super(...arguments);
        this.fsp = this.config.fsp;
    }
    getSQLType() {
        const precision = this.fsp === undefined ? '' : `(${this.fsp})`;
        return `timestamp${precision}`;
    }
    mapFromDriverValue(value) {
        return new Date(value + '+0000');
    }
    mapToDriverValue(value) {
        return value.toISOString().slice(0, 19).replace('T', ' ');
    }
}
_b$6 = alias$1.entityKind;
MySqlTimestamp[_b$6] = 'MySqlTimestamp';
class MySqlTimestampStringBuilder extends MySqlDateColumnBaseBuilder {
    constructor(name, config) {
        super(name);
        this.config.fsp = config?.fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlTimestampString(table, this.config);
    }
}
_c$1 = alias$1.entityKind;
MySqlTimestampStringBuilder[_c$1] = 'MySqlTimestampStringBuilder';
class MySqlTimestampString extends MySqlDateBaseColumn {
    constructor() {
        super(...arguments);
        this.fsp = this.config.fsp;
    }
    getSQLType() {
        const precision = this.fsp === undefined ? '' : `(${this.fsp})`;
        return `timestamp${precision}`;
    }
}
_d = alias$1.entityKind;
MySqlTimestampString[_d] = 'MySqlTimestampString';
function timestamp(name, config = {}) {
    if (config.mode === 'string') {
        return new MySqlTimestampStringBuilder(name, config);
    }
    return new MySqlTimestampBuilder(name, config);
}

var _a$6, _b$5;
class MySqlTinyIntBuilder extends session.MySqlColumnBuilderWithAutoIncrement {
    /** @internal */
    build(table) {
        return new MySqlTinyInt(table, this.config);
    }
}
_a$6 = alias$1.entityKind;
MySqlTinyIntBuilder[_a$6] = 'MySqlTinyIntBuilder';
class MySqlTinyInt extends session.MySqlColumnWithAutoIncrement {
    getSQLType() {
        return 'tinyint';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'string') {
            return Number(value);
        }
        return value;
    }
}
_b$5 = alias$1.entityKind;
MySqlTinyInt[_b$5] = 'MySqlTinyInt';
function tinyint(name) {
    return new MySqlTinyIntBuilder(name);
}

var _a$5, _b$4;
class MySqlVarBinaryBuilder extends session.MySqlColumnBuilder {
    /** @internal */
    constructor(name, config) {
        super(name);
        this.config.length = config?.length;
    }
    /** @internal */
    build(table) {
        return new MySqlVarBinary(table, this.config);
    }
}
_a$5 = alias$1.entityKind;
MySqlVarBinaryBuilder[_a$5] = 'MySqlVarBinaryBuilder';
class MySqlVarBinary extends session.MySqlColumn {
    constructor() {
        super(...arguments);
        this.length = this.config.length;
    }
    getSQLType() {
        return this.length === undefined ? `varbinary` : `varbinary(${this.length})`;
    }
}
_b$4 = alias$1.entityKind;
MySqlVarBinary[_b$4] = 'MySqlVarBinary';
function varbinary(name, options) {
    return new MySqlVarBinaryBuilder(name, options);
}

var _a$4, _b$3;
class MySqlVarCharBuilder extends session.MySqlColumnBuilder {
    /** @internal */
    constructor(name, config) {
        super(name);
        this.config.length = config.length;
        this.config.enum = config.enum;
    }
    /** @internal */
    build(table) {
        return new MySqlVarChar(table, this.config);
    }
}
_a$4 = alias$1.entityKind;
MySqlVarCharBuilder[_a$4] = 'MySqlVarCharBuilder';
class MySqlVarChar extends session.MySqlColumn {
    constructor() {
        super(...arguments);
        this.length = this.config.length;
        this.enumValues = (this.config.enum ?? []);
    }
    getSQLType() {
        return this.length === undefined ? `varchar` : `varchar(${this.length})`;
    }
}
_b$3 = alias$1.entityKind;
MySqlVarChar[_b$3] = 'MySqlVarChar';
function varchar(name, config) {
    return new MySqlVarCharBuilder(name, config);
}

var _a$3, _b$2;
class MySqlYearBuilder extends session.MySqlColumnBuilder {
    /** @internal */
    build(table) {
        return new MySqlYear(table, this.config);
    }
}
_a$3 = alias$1.entityKind;
MySqlYearBuilder[_a$3] = 'MySqlYearBuilder';
class MySqlYear extends session.MySqlColumn {
    getSQLType() {
        return `year`;
    }
}
_b$2 = alias$1.entityKind;
MySqlYear[_b$2] = 'MySqlYear';
function year(name) {
    return new MySqlYearBuilder(name);
}

var _a$2, _b$1, _c;
class IndexBuilderOn {
    constructor(name, unique) {
        this.name = name;
        this.unique = unique;
    }
    on(...columns) {
        return new IndexBuilder(this.name, columns, this.unique);
    }
}
_a$2 = alias$1.entityKind;
IndexBuilderOn[_a$2] = 'MySqlIndexBuilderOn';
class IndexBuilder {
    constructor(name, columns, unique) {
        this.config = {
            name,
            columns,
            unique,
        };
    }
    using(using) {
        this.config.using = using;
        return this;
    }
    algorythm(algorythm) {
        this.config.algorythm = algorythm;
        return this;
    }
    lock(lock) {
        this.config.lock = lock;
        return this;
    }
    /** @internal */
    build(table) {
        return new Index(this.config, table);
    }
}
_b$1 = alias$1.entityKind;
IndexBuilder[_b$1] = 'MySqlIndexBuilder';
class Index {
    constructor(config, table) {
        this.config = { ...config, table };
    }
}
_c = alias$1.entityKind;
Index[_c] = 'MySqlIndex';
function index(name) {
    return new IndexBuilderOn(name, false);
}
function uniqueIndex(name) {
    return new IndexBuilderOn(name, true);
}

var _a$1, _b;
function primaryKey(...columns) {
    return new PrimaryKeyBuilder(columns);
}
class PrimaryKeyBuilder {
    constructor(columns) {
        this.columns = columns;
    }
    /** @internal */
    build(table) {
        return new PrimaryKey(table, this.columns);
    }
}
_a$1 = alias$1.entityKind;
PrimaryKeyBuilder[_a$1] = 'MySqlPrimaryKeyBuilder';
class PrimaryKey {
    constructor(table, columns) {
        this.table = table;
        this.columns = columns;
    }
    getName() {
        return `${this.table[session.MySqlTable.Symbol.Name]}_${this.columns.map((column) => column.name).join('_')}_pk`;
    }
}
_b = alias$1.entityKind;
PrimaryKey[_b] = 'MySqlPrimaryKey';

var _a;
class MySqlSchema {
    constructor(schemaName) {
        this.schemaName = schemaName;
        this.table = (name, columns, extraConfig) => {
            return session.mysqlTableWithSchema(name, columns, extraConfig, this.schemaName);
        };
        this.view = ((name, columns) => {
            return session.mysqlViewWithSchema(name, columns, this.schemaName);
        });
    }
}
_a = alias$1.entityKind;
MySqlSchema[_a] = 'MySqlSchema';
/** @deprecated - use `instanceof MySqlSchema` */
function isMySqlSchema(obj) {
    return alias$1.is(obj, MySqlSchema);
}
/**
 * Create a MySQL schema.
 * https://dev.mysql.com/doc/refman/8.0/en/create-database.html
 *
 * @param name mysql use schema name
 * @returns MySQL schema
 */
function mysqlDatabase(name) {
    return new MySqlSchema(name);
}
/**
 * @see mysqlDatabase
 */
const mysqlSchema = mysqlDatabase;

function getTableConfig(table) {
    const columns = Object.values(table[session.MySqlTable.Symbol.Columns]);
    const indexes = [];
    const checks = [];
    const primaryKeys = [];
    const uniqueConstraints = [];
    const foreignKeys = Object.values(table[session.MySqlTable.Symbol.InlineForeignKeys]);
    const name = table[alias$1.Table.Symbol.Name];
    const schema = table[alias$1.Table.Symbol.Schema];
    const baseName = table[alias$1.Table.Symbol.BaseName];
    const extraConfigBuilder = table[session.MySqlTable.Symbol.ExtraConfigBuilder];
    if (extraConfigBuilder !== undefined) {
        const extraConfig = extraConfigBuilder(table[session.MySqlTable.Symbol.Columns]);
        for (const builder of Object.values(extraConfig)) {
            if (alias$1.is(builder, IndexBuilder)) {
                indexes.push(builder.build(table));
            }
            else if (alias$1.is(builder, CheckBuilder)) {
                checks.push(builder.build(table));
            }
            else if (alias$1.is(builder, session.UniqueConstraintBuilder)) {
                uniqueConstraints.push(builder.build(table));
            }
            else if (alias$1.is(builder, PrimaryKeyBuilder)) {
                primaryKeys.push(builder.build(table));
            }
            else if (alias$1.is(builder, session.ForeignKeyBuilder)) {
                foreignKeys.push(builder.build(table));
            }
        }
    }
    return {
        columns,
        indexes,
        foreignKeys,
        checks,
        primaryKeys,
        uniqueConstraints,
        name,
        schema,
        baseName,
    };
}
function getViewConfig(view) {
    return {
        ...view[alias$1.ViewBaseConfig],
        ...view[session.MySqlViewConfig],
    };
}

exports.ForeignKey = session.ForeignKey;
exports.ForeignKeyBuilder = session.ForeignKeyBuilder;
exports.InlineForeignKeys = session.InlineForeignKeys;
exports.ManualViewBuilder = session.ManualViewBuilder;
exports.MySqlColumn = session.MySqlColumn;
exports.MySqlColumnBuilder = session.MySqlColumnBuilder;
exports.MySqlColumnBuilderWithAutoIncrement = session.MySqlColumnBuilderWithAutoIncrement;
exports.MySqlColumnWithAutoIncrement = session.MySqlColumnWithAutoIncrement;
exports.MySqlDatabase = session.MySqlDatabase;
exports.MySqlDelete = session.MySqlDelete;
exports.MySqlDialect = session.MySqlDialect;
exports.MySqlInsert = session.MySqlInsert;
exports.MySqlInsertBuilder = session.MySqlInsertBuilder;
exports.MySqlSelect = session.MySqlSelect;
exports.MySqlSelectBuilder = session.MySqlSelectBuilder;
exports.MySqlSelectQueryBuilder = session.MySqlSelectQueryBuilder;
exports.MySqlSession = session.MySqlSession;
exports.MySqlTable = session.MySqlTable;
exports.MySqlTransaction = session.MySqlTransaction;
exports.MySqlUpdate = session.MySqlUpdate;
exports.MySqlUpdateBuilder = session.MySqlUpdateBuilder;
exports.MySqlView = session.MySqlView;
exports.MySqlViewBase = session.MySqlViewBase;
exports.MySqlViewConfig = session.MySqlViewConfig;
exports.PreparedQuery = session.PreparedQuery;
exports.QueryBuilder = session.QueryBuilder;
exports.UniqueConstraint = session.UniqueConstraint;
exports.UniqueConstraintBuilder = session.UniqueConstraintBuilder;
exports.UniqueOnConstraintBuilder = session.UniqueOnConstraintBuilder;
exports.ViewBuilder = session.ViewBuilder;
exports.ViewBuilderCore = session.ViewBuilderCore;
exports.foreignKey = session.foreignKey;
exports.mysqlTable = session.mysqlTable;
exports.mysqlTableCreator = session.mysqlTableCreator;
exports.mysqlTableWithSchema = session.mysqlTableWithSchema;
exports.mysqlView = session.mysqlView;
exports.mysqlViewWithSchema = session.mysqlViewWithSchema;
exports.unique = session.unique;
exports.uniqueKeyName = session.uniqueKeyName;
exports.Check = Check;
exports.CheckBuilder = CheckBuilder;
exports.Index = Index;
exports.IndexBuilder = IndexBuilder;
exports.IndexBuilderOn = IndexBuilderOn;
exports.MySqlBigInt53 = MySqlBigInt53;
exports.MySqlBigInt53Builder = MySqlBigInt53Builder;
exports.MySqlBigInt64 = MySqlBigInt64;
exports.MySqlBigInt64Builder = MySqlBigInt64Builder;
exports.MySqlBinary = MySqlBinary;
exports.MySqlBinaryBuilder = MySqlBinaryBuilder;
exports.MySqlBoolean = MySqlBoolean;
exports.MySqlBooleanBuilder = MySqlBooleanBuilder;
exports.MySqlChar = MySqlChar;
exports.MySqlCharBuilder = MySqlCharBuilder;
exports.MySqlCustomColumn = MySqlCustomColumn;
exports.MySqlCustomColumnBuilder = MySqlCustomColumnBuilder;
exports.MySqlDate = MySqlDate;
exports.MySqlDateBuilder = MySqlDateBuilder;
exports.MySqlDateString = MySqlDateString;
exports.MySqlDateStringBuilder = MySqlDateStringBuilder;
exports.MySqlDateTime = MySqlDateTime;
exports.MySqlDateTimeBuilder = MySqlDateTimeBuilder;
exports.MySqlDateTimeString = MySqlDateTimeString;
exports.MySqlDateTimeStringBuilder = MySqlDateTimeStringBuilder;
exports.MySqlDecimal = MySqlDecimal;
exports.MySqlDecimalBuilder = MySqlDecimalBuilder;
exports.MySqlDouble = MySqlDouble;
exports.MySqlDoubleBuilder = MySqlDoubleBuilder;
exports.MySqlEnumColumn = MySqlEnumColumn;
exports.MySqlEnumColumnBuilder = MySqlEnumColumnBuilder;
exports.MySqlFloat = MySqlFloat;
exports.MySqlFloatBuilder = MySqlFloatBuilder;
exports.MySqlInt = MySqlInt;
exports.MySqlIntBuilder = MySqlIntBuilder;
exports.MySqlJson = MySqlJson;
exports.MySqlJsonBuilder = MySqlJsonBuilder;
exports.MySqlMediumInt = MySqlMediumInt;
exports.MySqlMediumIntBuilder = MySqlMediumIntBuilder;
exports.MySqlReal = MySqlReal;
exports.MySqlRealBuilder = MySqlRealBuilder;
exports.MySqlSchema = MySqlSchema;
exports.MySqlSerial = MySqlSerial;
exports.MySqlSerialBuilder = MySqlSerialBuilder;
exports.MySqlSmallInt = MySqlSmallInt;
exports.MySqlSmallIntBuilder = MySqlSmallIntBuilder;
exports.MySqlText = MySqlText;
exports.MySqlTextBuilder = MySqlTextBuilder;
exports.MySqlTime = MySqlTime;
exports.MySqlTimeBuilder = MySqlTimeBuilder;
exports.MySqlTimestamp = MySqlTimestamp;
exports.MySqlTimestampBuilder = MySqlTimestampBuilder;
exports.MySqlTimestampString = MySqlTimestampString;
exports.MySqlTimestampStringBuilder = MySqlTimestampStringBuilder;
exports.MySqlTinyInt = MySqlTinyInt;
exports.MySqlTinyIntBuilder = MySqlTinyIntBuilder;
exports.MySqlVarBinary = MySqlVarBinary;
exports.MySqlVarBinaryBuilder = MySqlVarBinaryBuilder;
exports.MySqlVarChar = MySqlVarChar;
exports.MySqlVarCharBuilder = MySqlVarCharBuilder;
exports.MySqlYear = MySqlYear;
exports.MySqlYearBuilder = MySqlYearBuilder;
exports.PrimaryKey = PrimaryKey;
exports.PrimaryKeyBuilder = PrimaryKeyBuilder;
exports.alias = alias;
exports.bigint = bigint;
exports.binary = binary;
exports.boolean = boolean;
exports.char = char;
exports.check = check;
exports.customType = customType;
exports.date = date;
exports.datetime = datetime;
exports.decimal = decimal;
exports.double = double;
exports.float = float;
exports.getTableConfig = getTableConfig;
exports.getViewConfig = getViewConfig;
exports.index = index;
exports.int = int;
exports.isMySqlSchema = isMySqlSchema;
exports.json = json;
exports.longtext = longtext;
exports.mediumint = mediumint;
exports.mediumtext = mediumtext;
exports.mysqlDatabase = mysqlDatabase;
exports.mysqlEnum = mysqlEnum;
exports.mysqlSchema = mysqlSchema;
exports.primaryKey = primaryKey;
exports.real = real;
exports.serial = serial;
exports.smallint = smallint;
exports.text = text;
exports.time = time;
exports.timestamp = timestamp;
exports.tinyint = tinyint;
exports.tinytext = tinytext;
exports.uniqueIndex = uniqueIndex;
exports.varbinary = varbinary;
exports.varchar = varchar;
exports.year = year;
